File: app/lib/content/outline.dart
import 'book.dart';
import 'fa_numbers.dart';

class OutlineSection {
  const OutlineSection(this.title, this.pdfPage, this.printedPage);

  final String title;
  final int pdfPage;
  final int? printedPage;
}

class OutlineChapter {
  OutlineChapter(this.chapter, this.pdfPage);

  final Chapter chapter;
  final int pdfPage;
  final List<OutlineSection> sections = [];
}

const _teachingKinds = {'content', 'exercises', 'review'};

/// The book's chapters and sections in reading order, built from the page packages.
/// Needs [Book.preloadPackages] to have run.
List<OutlineChapter> buildOutline(Book book) {
  final chapters = <OutlineChapter>[];
  for (var page = 1; page <= book.pageCount; page++) {
    final pkg = book.cachedPackage(page);
    final chapter = book.chapter(pkg?.chapter);
    if (pkg == null || chapter == null) continue;
    if (chapters.isEmpty || chapters.last.chapter.number != chapter.number) {
      chapters.add(OutlineChapter(chapter, page));
    }
    final title = pkg.sectionFa?.trim();
    if (title == null || title.isEmpty || !_teachingKinds.contains(pkg.json['page_kind'])) continue;
    final sections = chapters.last.sections;
    if (sections.isNotEmpty && sections.last.title == title) continue;
    sections.add(OutlineSection(title, page, pkg.printedPage));
  }
  return chapters;
}

/// "[فهرست کتاب]" block for the teacher's prompt. Deterministic, so it stays in the prompt cache.
String outlineText(Book book, List<OutlineChapter> outline) {
  final lines = ['[فهرست کتاب: ${book.titleFa}]'];
  for (final c in outline) {
    final ch = c.chapter;
    lines.add('فصل ${faNum(ch.number)}: ${ch.titleFa} (صفحات ${faNum(ch.printedStart)} تا ${faNum(ch.printedEnd)})');
    for (final s in c.sections) {
      lines.add(s.printedPage == null ? '- ${s.title}' : '- ${s.title} (صفحهٔ ${faNum(s.printedPage!)})');
    }
  }
  return lines.join('\n');
}

File: UI Code/src/tutor/outline.ts
import { faNum } from "./fa";
import type { Chapter, TextBook } from "./book";

export type OutlineSection = { title: string; pdfPage: number; printedPage: number | null };
export type OutlineChapter = { chapter: Chapter; pdfPage: number; sections: OutlineSection[] };

export const TEACHING_KINDS = new Set(["content", "exercises", "review"]);

/** The book's chapters and sections in reading order, built from the page packages. */
export function buildOutline(book: TextBook): OutlineChapter[] {
  const chapters: OutlineChapter[] = [];
  for (let page = 1; page <= book.pageCount; page++) {
    const pkg = book.pkg(page);
    const chapter = book.chapterOfPage(page);
    if (!chapter) continue;
    if (!chapters.length || chapters[chapters.length - 1].chapter.number !== chapter.number) {
      chapters.push({ chapter, pdfPage: page, sections: [] });
    }
    const title = pkg?.sectionFa?.trim();
    if (!pkg || !title || !TEACHING_KINDS.has(pkg.pageKind)) continue;
    const sections = chapters[chapters.length - 1].sections;
    if (sections.length && sections[sections.length - 1].title === title) continue;
    sections.push({ title, pdfPage: page, printedPage: pkg.printedPage });
  }
  return chapters;
}

/** "[فهرست کتاب]" block for the teacher's prompt. Deterministic, so it stays in the prompt cache. */
export function outlineText(book: TextBook, outline: OutlineChapter[]): string {
  const lines = [`[فهرست کتاب: ${book.titleFa}]`];
  for (const c of outline) {
    const ch = c.chapter;
    lines.push(`فصل ${faNum(ch.number)}: ${ch.titleFa} (صفحات ${faNum(ch.printedStart)} تا ${faNum(ch.printedEnd)})`);
    for (const s of c.sections) {
      lines.push(s.printedPage === null ? `- ${s.title}` : `- ${s.title} (صفحهٔ ${faNum(s.printedPage)})`);
    }
  }
  return lines.join("\n");
}

/** "خواص لیمت (Properties of limit)" → "خواص لیمت"; parentheses holding a formula stay. */
export const stripLatin = (s: string) => s.replace(/\s*\((?=[^)]*[A-Za-z]{2})[^)$\\]*\)/g, "").trim();
